pkg/providers/amifamily/bootstrap/custom.go

```diff
@@ -15,15 +15,47 @@ limitations under the License.
 package bootstrap
 
 import (
+	"bytes"
 	"encoding/base64"
-
-	"github.com/aws/aws-sdk-go/aws"
+	"github.com/samber/lo"
+	v1 "k8s.io/api/core/v1"
+	"text/template"
 )
 
 type Custom struct {
 	Options
 }
 
+type TemplateData struct {
```

Comment:
This is really cool! I'm curious on a couple of things:
1. Do you think that this kind of templating might be useful for other AMIFamilies within Karpenter?
2. Karpenter currently has a "node registration flow" which adds all of the labels that Karpenter is tracking on its Machine CustomResource onto the Node when the Node joins the cluster. Is this mechanism sufficient to get the labels onto the node or do you also need these labels passed through into the userData?

Reply:
It could be useful for sure. It all depends on how each registers though. I’m not as familiar with the others.

Reply:
Looking at like bottlerocket, it wouldn't be useful because it's already passing in taints and labels via it's special toml configuration https://github.com/aws/karpenter/blob/main/pkg/providers/amifamily/bootstrap/bottlerocket.go#L47-L49

So I feel it's likely to be implementation specific.

Reply:
How do you think this templating plays with the ideas presented here: https://github.com/aws/karpenter/issues/4459? One thought that we had was to drop a file onto the filesystem that you could modify and pull data from rather than having to template the userData like this.

Reply:
I think that's a different problem entirely as it assumes EKS. This simple just makes the data available to the custom user data. At least with the custom user data it's expected to be passed through the CRD definition. For the others the file already exists or is generated it seems depending on the family.

Reply:
On first glance, this is a brilliant idea imo. It gives the user access to karpenters state (runtime interface) at launch time, without assuming anything about the file system. It's even user data format agnostic, i.e. could work with bash or toml. 

To Jonathan's point, I want us to be very careful about what exactly we expose in this template, as it will become something we support forever. 

Specifically on labels and taints, we've been hoping to shift label and taints registration to the controller, instead of the bootstrap. Does that mean you should be able to know the labels at bootstrap time? I don't know -- maybe.

Reply:
To Ellis's point @ekristen, do you need these taints and labels added in the userData script if they are added to the Node at node-join time when the the node first starts up?

Reply:
Ah. I guess in my initial testing this wasn't happening but for different reasons at least labels work. I'll need to verify taints.

I don't use cloud provider operators so I had to write a shim to get the cloud provider id to be set so it would link up to karpenter.

Reply:
> I don't use cloud provider operators so I had to write a shim to get the cloud provider id to be set so it would link up to karpenter. 

I could still see this being _extremely_ useful for things like the kubeletConfiguration if you are trying to inject the kubelet args that Karpenter is aware of into your userData script. I think, in general, I'm supportive of this templating mechanism since we can expand it out to other areas of Karpenter configuration that might affect userData.